### DISCUSSION TITLE
'mini.files' + LSP

### DISCUSSION THREAD

**Discussion by echasnovski** [nvim-mini/mini.nvim#400] [Ideas]
Right now, 'mini.files' has [events](https://github.com/echasnovski/mini.nvim/blob/bb8ef7cfaf7b0c4492836f318df0b61e92ea3de1/doc/mini-files.txt#L331) triggered after successful file system action is performed. It seems to me like the best design to allow any user-specific actions which should follow file system manipulation. In particular, LSP actions via `DidCreateFiles`/`DidRenameFiles`/`DidDeleteFiles`.

If anybody has ideas/snippets/results of how this can be implements, lets discuss them here.

Relevant discussions: #377 (in particular, [this comment](https://github.com/echasnovski/mini.nvim/issues/377#issuecomment-1620011346) and [this comment](https://github.com/echasnovski/mini.nvim/issues/377#issuecomment-1620047456)), #378.

**Comment by echasnovski** [comment]
Answer to [this comment](https://github.com/echasnovski/mini.nvim/issues/377#issuecomment-1620069586): 

> you already list all planned actions for the user to confirm, which may be wrong? So what's wrong with having that programmatically. Just like the other problems with unexpected fs actions mentioned above, the user can just synchronize more often right?

Those are indeed only planned actions which 'mini.files' will **try** to do.

More frequent synchronization will not solve all problems: moving directory inside itslef is not possible, manipulating files/directories inside directory without write permissions, etc.

> As it is, the functionality of lsp refactoring is completely impossible, with the pre action callback it becomes possible but with some edge cases that you can just let the user deal with in the callback that they are writing. All that is needed on mini.files side is the callback with the information, a couple of lines of code. I don't think it's a concern/responsibility of the plugin that the user could make mistakes with the callback. This is always a problem with file explorers and their operations, but atleast in this case you don't have to think about it yourself just let the user worry about it.

I don't really understand, is `DidCreateFiles`/`DidRenameFiles`/`DidDeleteFiles` not enough for Lsp refactoring?

**Reply by IndianBoy42** [reply]
No, the `did*Files` notifications are just for notifying the server that some fs operation was done. It is a notification not a request, so there is no response to the client. It is only if the server/needs or wants it to update some internal state itself. 

the `will*Files` methods are where LSPs return `workspaceEdits`, in the same way as code actions or variable rename requests, that the client should apply *before* doing the fs operation, according to the specification. Thats the spec, so all servers will do their computations and return edits assuming none of the fs operations are done yet. 

> Those are indeed only planned actions which 'mini.files' will try to do.

Right, so whats the difference with the user seeing that and assuming they will be done, versus the callback (user written code) seeing that and assuming they will be done? If there is a mistake there is a mistake, letting the user be in control of it (that is including the users code). The user can just undo any edits. Perhaps an on failure event might be useful if we want to rollback the workspace edits, but I don't know if any other editor does that.

**Reply by IndianBoy42** [reply]
Als, we should focus on when/whether file creation, deletion or renaming can fail. Moving files seems harder to detect possible failures, but that is not a lsp request actually. (regardless if the user wants to do things based on the files moving then its their responsibility and I think thats fine for a plugin that wants to stay as a minimal and extensible core)

**Reply by echasnovski** [reply]
> No, the `did*Files` notifications are just for notifying the server that some fs operation was done. It is a notification not a request, so there is no response to the client. It is only if the server/needs or wants it to update some internal state itself.
> 
> the `will*Files` methods are where LSPs return `workspaceEdits`, in the same way as code actions or variable rename requests, that the client should apply _before_ doing the fs operation, according to the specification. Thats the spec, so all servers will do their computations and return edits assuming none of the fs operations are done yet.

That's... a very *special* design choice. I can't think of any its benefit compared to reverse design (`did*Files` perform actions, `will*Files` is a notification that something will/might happen).

> > Those are indeed only planned actions which 'mini.files' will try to do.
> 
> Right, so whats the difference with the user seeing that and assuming they will be done, versus the callback (user written code) seeing that and assuming they will be done? If there is a mistake there is a mistake, letting the user be in control of it (that is including the users code). The user can just undo any edits. Perhaps an on failure event might be useful if we want to rollback the workspace edits, but I don't know if any other editor does that.

The difference is that with user interaction they can see right away what was actually done. In the callback there won't be such a feedback and will result essentially result into workspace edits on an action which wasn't actually done. And those edits can be hard to track, as far as I understand.

> Als, we should focus on when/whether file creation, deletion or renaming can fail. Moving files seems harder to detect possible failures, but that is not a lsp request actually. (regardless if the user wants to do things based on the files moving then its their responsibility and I think thats fine for a plugin that wants to stay as a minimal and extensible core)

I think actually trying to use present post-events to make `will*Files` requests is best here. I don't think it should fail miserably, at least. For example, I've looked at rust-lang/rust-analyzer#7009 and it seems to compute workspace edits only based on request information.

**Reply by IndianBoy42** [reply]
> The difference is that with user interaction they can see right away what was actually done. In the callback there won't be such a feedback and will result essentially result into workspace edits on an action which wasn't actually done. And those edits can be hard to track, as far as I understand.

Looking at every other lsp client implementation I can find this isn't a concern anyone else seems to worry about. Apart from some basic sanity checks to the fs operation.

> That's... a very special design choice. I can't think of any its benefit compared to reverse design (did*Files perform actions, will*Files is a notification that something will/might happen).

Having willRenameFiles return the edit allows the editor to preview the refactoring changes before making any changes. All the `did*` notifications are just to keep the server in sync with the filesystem state (since the protocol is designed to work even if the server doesn't have direct access to the client filesystem, although these notifications are actually not necessary in most cases)

> I think actually trying to use present post-events to make will*Files requests is best here. I don't think it should fail miserably, at least. 

I'd obviously rather follow the spec rather than pray that everything makes sense when done out of order. That will be even more invisible and hard to detect as a failure. Also since they are autocmds and asynchronous then applying workspaceEdits from here seems like a bad idea too. Again not sure about any of this because its far out of how things are 'supposed' to be done and the concerns are all in edge cases so it would be extremely hard to test. Going with the way that every other editor does it is the simplest way.

> The difference is that with user interaction they can see right away what was actually done. In the callback there won't be such a feedback and will result essentially result into workspace edits on an action which wasn't actually done. And those edits can be hard to track, as far as I understand.

I still think letting users have the freedom to customize (and make mistakes) is valuable. Obviously we can never predict what will happen when it comes to fs operations, but version control, undo history and backups exist. None of this is fatal or so precious that we have to protect the user from themselves. I understand if its a big coding/maintenance burden that not every customization point can be added, but in this case just one callback covers the major use case. 

If the pre operation callbacks are early enough (pre confirmation) then the callback can pop up some floating windows with previews of the edits.

**Reply by echasnovski** [reply]
> Looking at every other lsp client implementation I can find this isn't a concern anyone else seems to worry about. Apart from some basic sanity checks to the fs operation.

This is exactly my concern, because they seem to rely on the fact that `will*Files` are emitted prior to actions which are **assumed to succeed**. 'mini.files' won't give such guarantees. 

> Having willRenameFiles return the edit allows the editor to preview the refactoring changes before making any changes. 

I see. This does make *some* sense, albeit slightly confusing: why this can't be done in `did*Files`? I understand that this is the specification right now and more confused towards its design decisions.

> I still think letting users have the freedom to customize (and make mistakes) is valuable. Obviously we can never predict what will happen when it comes to fs operations, but version control, undo history and backups exist. None of this is fatal or so precious that we have to protect the user from themselves. I understand if its a big coding/maintenance burden that not every customization point can be added, but in this case just one callback covers the major use case.
> 
> If the pre operation callbacks are early enough (pre confirmation) then the callback can pop up some floating windows with previews of the edits.

The only possible design I consider here is to trigger event prior executing all actions, meaning *after* confirmation.

------

Let's take this into more grounded conversation.

I won't implement this LSP functionality. If you want to try doing it, tell me what you might need from code and I'll help you making proper change to 'mini.files' source. Once you implement something you are happy with, we'll potentially iterate about the best way to merge those changes into 'mini.files' and celebrate the new functionality.

Right now I understand that single event with all future **attempted** file system actions should be enough, right? If so, add the following line as the first command of [`H.fs_actions_apply()`](https://github.com/echasnovski/mini.nvim/blob/bb8ef7cfaf7b0c4492836f318df0b61e92ea3de1/lua/mini/files.lua#L2293):

```lua
  H.trigger_event('MiniFilesActionPre', { action = 'pre', fs_actions = fs_actions })
```

This will trigger event prior to attempting other actions. `fs_actions` is a table with field for every future action. `create` and `delete` are arrays of paths, others - arrays of tables with `from` and `to` fields.

Does this sound good?

**Reply by NU11V01D** [reply]
Did this get any follow up afterwards? I doubt this went stale after nearly three years from being posted. Was there any release notes or note in the help file that I missed? I am truly looking forward to implementing this feature the `mini` way; I love the plugin! I'd gladly contribute to this and test it myself, but sadly I don't have that much experience with Lua yet.

**Reply by Paulo4lvarez** [reply]
> Right now I understand that single event with all future attempted file system actions should be enough, right? If so, add the following line as the first command of [H.fs_actions_apply()](https://github.com/echasnovski/mini.nvim/blob/bb8ef7cfaf7b0c4492836f318df0b61e92ea3de1/lua/mini/files.lua#L2293):

Would this implementation still work on the current version of the plugin? If it does or if it isn't necessary because it was already added then something in the examples page would be useful.

**Reply by echasnovski** [reply]
> Did this get any follow up afterwards? I doubt this went stale after nearly three years from being posted. Was there any release notes or note in the help file that I missed? I am truly looking forward to implementing this feature the `mini` way; I love the plugin! I'd gladly contribute to this and test it myself, but sadly I don't have that much experience with Lua yet.

No, this is not stale: having built-in LSP support for 'mini.files' is planned. It has certain implementation issues (see #1586 comments) which need a research as to how popular LSP servers behave and somewhat low-ish priority for me. But I do have plans to tackle it after the current "big project" is finished.

And it has not yet been two years since 'mini.files' release, let alone three years after starting this discussion.

**Reply by Paulo4lvarez** [reply]
Alright, thanks for the reply! Good to know it is planned. Your plugins are amazing!

**Comment by stuckinsnow** [comment]
In case this is useful to anyone, I added these to my config: 

```
      vim.api.nvim_create_autocmd("User", {
        pattern = "MiniFilesActionRename",
        callback = function(event)
          Snacks.rename.on_rename_file(event.data.from, event.data.to)
        end,
      })

      vim.api.nvim_create_autocmd("User", {
        pattern = "MiniFilesActionMove",
        callback = function(event)
          Snacks.rename.on_rename_file(event.data.from, event.data.to)
        end,
      })
```

You can also do it with the code below:


```
      vim.api.nvim_create_autocmd("User", {
        pattern = "MiniFilesActionRename",
        callback = function(event)
          Snacks.rename.on_rename_file(event.data.from, event.data.to)
        end,
      })

      vim.api.nvim_create_autocmd("User", {
        desc = "Update imports when files are moved",
        pattern = "MiniFilesActionMove",
        callback = function(event)
          local old_path = event.data.from
          local new_path = event.data.to

          vim.schedule(function()
            -- Get all active LSP clients (compatible with newer Neovim versions)
            local clients = vim.lsp.get_clients()

            for _, client in pairs(clients) do
              -- Check if the client supports willRenameFiles
              if
                client.server_capabilities
                and client.server_capabilities.workspace
                and client.server_capabilities.workspace.fileOperations
                and client.server_capabilities.workspace.fileOperations.willRename
              then
                local params = {
                  files = {
                    {
                      oldUri = vim.uri_from_fname(old_path),
                      newUri = vim.uri_from_fname(new_path),
                    },
                  },
                }

                -- Send the willRenameFiles request to update imports
                client:request("workspace/willRenameFiles", params, function(err, workspace_edit)
                  if err then
                    vim.notify("LSP rename failed: " .. tostring(err), vim.log.levels.ERROR)
                    return
                  end

                  if workspace_edit then
                    -- Apply the workspace edit to update imports
                    vim.lsp.util.apply_workspace_edit(workspace_edit, client.offset_encoding)
                    vim.notify("Updated imports for moved file", vim.log.levels.INFO)
                  end
                end)
              end
            end
          end)
        end,
      })
```

**Comment by echasnovski** [comment]
LSP integration is now [part of the `main` branch](https://github.com/nvim-mini/mini.nvim/blob/4f6f84a96b076747b736d100388a8d7844771d7d/doc/mini-files.txt#L320-L332).
